### ISSUE TITLE
Suggestion: ghost MY in palette if inherited attribute selected in dropdown

### ISSUE THREAD

**Issue by brianharvey** [jmoenig/Snap#2327]
If you inherit X POSITION, the corresponding block is ghosted in the child.

But there's no WARDROBE block.  (Maybe there should be?  In Looks or in Sensing?)  So there's no obvious indication that you're inheriting costumes.  So, if, in the MY block in the palette, you set the dropdown to COSTUMES, I'm suggesting that the block should be ghosted.

Alternatively, add WARDROBE and JUKEBOX blocks.

**Comment by jmoenig** [comment]
But the stage can TELL the clone of a sorite to inherit sonething

Von meinem iPhone gesendet

**Comment by brianharvey** [comment]
?  But the issue isn't whether it can be inherited; the question is how to provide the user with a visual indication that it's inherited.  For variables, local methods, and _most_ attributes, a block in the child's palette is ghosted.  But there is no such block for wardrobe and jukebox; I'm suggesting that MY [COSTUMES] and MY [SOUNDS] be ghostable.  The user would still have to select those options to get the visible feedback, but at least it'd be possible.

**Comment by jmoenig** [comment]
Oops, my comment was meant to go to the other issue. Yeah, we eventually want better tools, you're right about that.

**Comment by jmoenig** [comment]
In the meantime I'm honestly struggling to even come up with compelling examples that demonstrate the benefits of inheritance in the first place. Sure, I do have a bunch of projects that basically show how this works, and then a few where inheritance is - arguably - beneficial because it centralizes shared properties. But ... really ... I couldn't so far come up with a GREAT example of why inheritance is cool in the first place. Actually, I'm kinda suspecting it's not :-/

**Comment by brianharvey** [comment]
Right, well, you're the one from an OOP background.  I'm a functional guy.

The main virtue of inheritance, for me, is debugging.  You can solve a problem in the parent, and you don't have to fix it again in the children. (And not just in code.  When I made the Mastermind project, I spent forever debugging the costumes.)  But secondly, you know, polymorphism.  A message to the child is delegated to the parent's method, which then calls a method in the child.

Really we need it because people (grownups, mainly, but some kids too) expect it in a serious programming language.

**Comment by bromagosa** [comment]
But proper OOP guys don't care that much about inheritance, they care about message passing ;)

Incidentally, though, I do use OOP to share code between sprites all the time and I find it very useful. It's great for games where you have [two paddles and two score counters](https://cloud.snap.berkeley.edu/snap/snap.html#present:Username=bromagosa&ProjectName=Pong&editMode&noRun), or [four Simon buttons](https://cloud.snap.berkeley.edu/snap/snap.html#present:Username=bromagosa&ProjectName=Simon&editMode&noRun), or [two motorbikes](https://cloud.snap.berkeley.edu/snap/snap.html#present:Username=bromagosa&ProjectName=Tron&editMode&noRun), etc.

**Comment by jmoenig** [comment]
Yes, you're both right. Bernat, now that you mention it I realize that I also probably unwittingly and automatically use OOP to share code between a group of similar sprites nearly all the time. Haha, it's almost become a habit that I don't even realize to be a special feature anymore. For example, for my keynote address in Klagenfurt this Thursday I made a simple ["Monty-Hall" project](https://snap.berkeley.edu/snapsource/dev/snap.html#present:Username=jens&ProjectName=Monty%20Hall), and, of course, the three doors all share the exact same code, and doors 2 and 3 inherit it from door 1.

But I'm not so sure whether we've found the right "habit" for this yes. Yes, code-sharing is - surprisingly - useful and even awesome to a degree I didn't anticipate in Snap. But I myself get bitten a lot when I make a change in one of the children instead of in the parent. Then I need to copy the added blocks over to the parent and reset the child to again inherit the code from the parent. Happens all the time to me, and that doesn't feel right. I've even been thinking whether perhaps John is right in insisting on classes instead of prototypes for GP.

Sometimes, actually, most of the times, my inheritance tree is just a single level. What I really want to do is work on three sprites that just share their code, regardless of which sprite is currently selected in the corral. It's been an enormous effort to make it easy for a sprite to "shadow" its scripts whenever you perform the slightest edit operation inside its scripting area. Maybe automatic "shadowing" has become too easy that way. Perhaps it would be better to *not* shadow by default among a "tribe" of sprites unless the user explicitly says so.

Oh well, I've got the hunch that this question has some interesting design possibilities for us to yet discover...

**Comment by brianharvey** [comment]
Maybe inherited scripts should be ghosted in the scripting area?  And maybe ghosted things need a context menu item "edit in parent" that switches to the right sprite.
